OPHD/Technology/Technology.h

```diff
@@ -1,18 +1,90 @@
 #pragma once
 
+#include <memory>
 #include <string>
 #include <vector>
 
-struct Effect;
 
 struct Technology
 {
-	std::string name;
-	std::string description;
+	struct Modifier;
+	struct Unlock;
+
+	struct Effect
+	{
+		enum class Type
+		{
+			Modifier,
+			Unlock
+		};
+
+		Effect() = delete;
+		virtual ~Effect() = default;
+
+		bool isModifier() const;
+		bool isUnlock() const;
+
+		Modifier& toModifier();
+		Unlock& toUnlock();
+
+	protected:
+		Effect(Type type);
+		
+		Type type;
+	};
+
+	struct Modifier : public Effect
+	{
+		enum class Modifies
+		{
+			AgricultureEfficiency,
+			BreakdownRate,
+			EducationEfficiency,
+			MaintenanceCost,
+			PopulationFertility,
+			PopulationMorale,
+			PopulationMortality,
+			RecyclingEfficiency,
+			SmelterEfficiency,
+			StructureCost,
+			StructureDecay
+		};
+
+		Modifier() = delete;
+		Modifier(Modifies modifies, float value);
+		~Modifier() override = default;
+
+		Modifies modifies;
+		float value;
+	};
+
+	struct Unlock : public Effect
+	{
+		enum class Unlocks
+		{
+			DisasterPrediction,
+			Robot,
+			Satellite,
+			Structure,
+			Vehicle
+		};
+
+		Unlock() = delete;
+		Unlock(Unlocks unlocks, std::string value);
+		~Unlock() override = default;
+
+		Unlocks unlocks;
+		std::string value;
+	};
+
+	std::string name{""};
+	std::string description{""};
 	
-	unsigned int id{0};
-	unsigned int labType{0};
-	unsigned int cost{0};
-	std::vector<unsigned int> requiredTechnologies;
-	std::vector<Effect> effects;
+	int id{0};
+	int labType{0};
+	int cost{0};
+
+	std::vector<int> requiredTechnologies;
+
+	std::vector<std::unique_ptr<Effect>> effects;

```

Comment:
This makes sense to me. I got used to C#'s boxing/unboxing semantics and sort of emulated that here. And since we're not adding/removing effects during runtime except when loading these can be separated and managed more easily.